examples/open_loop_control_fetch/run.py

```diff
@@ -13,92 +13,79 @@
 import spear
 import time
 
+# import observation_utils from common folder
+COMMON_DIR = os.path.realpath(os.path.join(os.path.dirname(__file__), ".."))
+import sys
+sys.path.append(COMMON_DIR)
+import common.observation_utils as observation_utils
+
+def get_action(action_names, action_values):
+    # len(action_values) - 1 as action_values contains an additional 'index' column
```

Comment:
- I think you can pass in `index=False` when calling `to_records()` to get rid of the index column, so then you don't need this extra logic.
- Use these functions.

```
def get_data_frame(action):
    columns = np.array([ [ name + ".x", name + ".y", name + ".z" ] for name, data in action.items() ] ]).ravel()   # append each name with .x, .y, .z
    data = [ [d] for d in np.array([ [ data[0], data[1], data[2] ] for name, data in action.items() ] ]).ravel() ] # get data as N*3 array of 1x1 arrays
    return pd.DataFrame(columns=columns, data=data)

def get_action(row):
    names = [ name[:-2] for name in row.dtype.names ][::3]                   # strip .x .y .z from each name, select every third entry
    data = np.array([ row[name] for name in row.dtype.names ]).reshape(-1,3) # get data as Nx3 array
    return dict(zip(names, data)
```

Reply:
Aah yes, I wanted to keep `index=True` as I was using `row["index"]` to get the row number. But, your suggestion is much cleaner, and I only need to add an extra row number tracker.

Reply:
Why is your `get_data_frame` function different from mine? Why is `data` a Python list with a single entry in it of type `np.array`?

```
def get_data_frame(action):
    columns = np.array([ [ name + ".x", name + ".y", name + ".z" ] for name, data in action.items() ]).ravel()  # append each name with .x, .y, .z
    data = [ np.array([ [ data[0], data[1], data[2] ] for name, data in action.items() ]).ravel() ]  # get data as list of unraveled N*3 array
    return pd.DataFrame(columns=columns, data=data)
```

Reply:
I think your understanding is right, i.e., `data` has to be a list of rows, where each row is a list/numpy array of values where each value corresponds to a column.

In my code, `data` is a single row entry, i.e., `data = [single row]`. In this single row, a numpy array, each array value corresponds to a column.
```bash
[SPEAR | generate_actions.py] [array([ 0.098,  0.   ,  0.   , -0.098,  0.   ,  0.   ,  0.   ,  0.   ,
        0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ,
        0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ,
        0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ,
        0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ])]
```

In your code, `data` is also a single row entry, i.e., `data = [single row]`. In this single row, a list, each value is a list.
```bash
[SPEAR | generate_actions.py] [[0.098], [0.0], [0.0], [-0.098], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0], [0.0]]
```

If I run your code, I get the following error
```bash
Traceback (most recent call last):
  File ".\generate_actions.py", line 48, in <module>
    df = pd.concat([df, get_data_frame(action)])
  File ".\generate_actions.py", line 32, in get_data_frame
    return pd.DataFrame(columns=columns, data=data)
  File "C:\Users\Rachith\miniconda3\envs\spear\lib\site-packages\pandas\core\frame.py", line 744, in __init__
    arrays, columns, index = nested_data_to_arrays(
  File "C:\Users\Rachith\miniconda3\envs\spear\lib\site-packages\pandas\core\internals\construction.py", line 510, in nested_data_to_arrays
    arrays, columns = to_arrays(data, columns, dtype=dtype)
  File "C:\Users\Rachith\miniconda3\envs\spear\lib\site-packages\pandas\core\internals\construction.py", line 875, in to_arrays
    content, columns = _finalize_columns_and_data(arr, columns, dtype)
  File "C:\Users\Rachith\miniconda3\envs\spear\lib\site-packages\pandas\core\internals\construction.py", line 972, in _finalize_columns_and_data
    raise ValueError(err) from err
ValueError: 39 columns passed, passed data had 1 columns
```

I haven't looked into why panda interprets your data as 1 column and not 39 columns.

Reply:
What do you get when you run the following code.

```
import numpy as np
import pandas as pd

# '1.5.1'
print(pd.__version__)

# array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])
A = np.arange(12)
print(A)

# ['col0', 'col1', 'col2', 'col3', 'col4', 'col5', 'col6', 'col7', 'col8', 'col9', 'col10', 'col11']
columns = [ "col" + str(a) for a in A ]
print(columns)

# array([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11]])
data = A.reshape(1, -1)
print(data)

#    col0  col1  col2  col3  col4  col5  col6  col7  col8  col9  col10  col11
# 0     0     1     2     3     4     5     6     7     8     9     10     11
df = pd.DataFrame(columns=columns, data=data)
print(df)
```

Reply:
```bash
# print(pd.__version__)
1.5.1

#print(A)
[ 0  1  2  3  4  5  6  7  8  9 10 11]

#print(columns)
['col0', 'col1', 'col2', 'col3', 'col4', 'col5', 'col6', 'col7', 'col8', 'col9', 'col10', 'col11']

#print(data)
[[ 0  1  2  3  4  5  6  7  8  9 10 11]]

#print(df)
   col0  col1  col2  col3  col4  col5  col6  col7  col8  col9  col10  col11
0     0     1     2     3     4     5     6     7     8     9     10     11
```

Reply:
If you're getting the same results as me, then why would this function be doing different things for us?

```
def get_data_frame(action):
    columns = np.array([ [name + ".x", name + ".y", name + ".z"] for name, _ in action.items() ]).ravel() # append .x .y .z to each action name
    data = np.array([ vec for _, vec in action.items() ]).reshape(1, -1)                                  # get data as array with shape (1, N*3)
    return pd.DataFrame(columns=columns, data=data)
```

In this code, `columns` is just an array of strings, and `data` is just an array of floats with the shape `(1, N*3)` where `N` is the number of entries in `action`. What is the difference between the standalone code snippet and the `get_data_frame` function above?

Reply:
I tried the following
```python
# [array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11])]
data_1 = [data.ravel()]
print(data_1)

#    col0  col1  col2  col3  col4  col5  col6  col7  col8  col9  col10  col11
# 0     0     1     2     3     4     5     6     7     8     9     10     11
df = pd.DataFrame(columns=columns, data=data_1)
print(df)

# [[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]]
data_2 = [[x for x in data.ravel()]]
print(data_2)

#    col0  col1  col2  col3  col4  col5  col6  col7  col8  col9  col10  col11
# 0     0     1     2     3     4     5     6     7     8     9     10     11
df = pd.DataFrame(columns=columns, data=data_2)
print(df)

# [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
data_3 = [x for x in data.ravel()]
print(data_3)

# error
df = pd.DataFrame(columns=columns, data=data_3)
print(df)

# [[0], [1], [2], [3], [4], [5], [6], [7], [8], [9], [10], [11]]
data_4 = [[x] for x in data.ravel()]
print(data_4)

# error
df = pd.DataFrame(columns=columns, data=data_4)
print(df)
```

So, modifying your line of code
```python
    data = [ [d] for d in np.array([ [data[0], data[1], data[2]] for _, data in action.items() ]).ravel() ]
```
to the below doesn't throw an error while creating pandas df.
```python
    data = [[ d for d in np.array([ [data[0], data[1], data[2]] for _, data in action.items() ]).ravel() ]]
or
    data = [[ [d] for d in np.array([ [data[0], data[1], data[2]] for _, data in action.items() ]).ravel() ]]
```

So, I think the issue was the outer `[]` braces which was missing.

Reply:
My revised `get_data_frame` function is simpler than any of those variants you tested.

```
data = np.array([ vec for _, vec in action.items() ]).reshape(1, -1)
```

This is simpler than any of the variants you tested because `reshape(1, -1)` is all that is required to obtain an iterable of iterables of floats.

Reply:
This is neat! 
Updated the code to use `reshape(1, -1)` in #259